prov/sm2/src/sm2_ep.c

```diff
@@ -296,6 +296,47 @@ static ssize_t sm2_do_inject(struct sm2_ep *ep, struct sm2_region *peer_smr,
 	return FI_SUCCESS;
 }
 
+static void sm2_format_cma(struct sm2_xfer_entry *xfer_entry,
+			   const struct iovec *iov, size_t count)
+{
+	struct sm2_cma_data *cma_data =
+		(struct sm2_cma_data *) xfer_entry->user_data;
+
+	/* All CMA messages must be delivery complete to stop completion from
+	 * being written instantly */
+	xfer_entry->hdr.op_flags |= FI_DELIVERY_COMPLETE;
+	xfer_entry->hdr.proto = sm2_proto_cma;
+	xfer_entry->hdr.size = ofi_total_iov_len(iov, count);
+	cma_data->iov_count = count;
+	memcpy(cma_data->iov, iov, sizeof(*iov) * count);
+}
+
+static ssize_t sm2_do_cma(struct sm2_ep *ep, struct sm2_region *peer_smr,
+			  sm2_gid_t peer_gid, uint32_t op, uint64_t tag,
+			  uint64_t data, uint64_t op_flags, struct ofi_mr **mr,
+			  const struct iovec *iov, size_t iov_count,
+			  size_t total_len, void *context)
+{
+	struct sm2_xfer_entry *xfer_entry;
+	struct sm2_av *av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_region *self_region = sm2_mmap_ep_region(&av->mmap, ep->gid);
+
+	if (smr_freestack_isempty(sm2_freestack(self_region))) {
+		sm2_progress_recv(ep);
```

Comment:
I remember you had this in your inject protocol as well. Remind me, is it for functional or performance reasons? I don't love the idea of the receive call driving receive progress. This is why we return EAGAIN and have the app do it.

Reply:
Functional... this is saying:
1. Check if you have available free stack entries
2. if no freestack entries are available, call recv (b/c someone might have returned one to you)
3. re-check if you have a freestack entry
4. If you do, great... send... if you don't return EAGAIN

This is required b/c some sender-only apps might never call progress.... and without this, they would never get their XFER entries back

Reply:
If a send/recv call returns EAGAIN, the app needs to call progress. It is part of the API definition.